src/imputors.jl

```diff
@@ -71,46 +87,49 @@ julia> impute(v, Interpolate())
  3.0
  4.0
  5.0
-```
-"""
-function impute(data, imp::Imputor; kwargs...)
-    # NOTE: We don't use a return type declaration here because `trycopy` isn't guaranteed
-    # to return the same type passed in. For example, subarrays and subdataframes will
-    # return a regular array or dataframe.
-    return impute!(trycopy(data), imp; kwargs...)
-end
-
-"""
-    impute!(data::A, imp; dims=:, kwargs...) -> A
-
-Impute the `missing` values in the array `data` using the imputor `imp`.
-Optionally, you can specify the dimension to impute along.
-
-# Arguments
-* `data::AbstractArray{Union{T, Missing}}`: the data to be impute along dimensions `dims`
-* `imp::Imputor`: the Imputor method to use
-
-# Keyword Arguments
-* `dims=:`: The dimension to impute along. `:rows` and `:cols` are also supported for matrices.
-
-# Returns
-* `AbstractArray{Union{T, Missing}}`: the input `data` with values imputed
-
-# NOTES
-1. Matrices have a deprecated `dims=2` special case as `dims=:` is a breaking change
-2. Mutation isn't guaranteed for all array types, hence we return the result
-3. `eachslice` is used internally which requires Julia 1.1
-
-# Example
-```jldoctest
-julia> using Impute: Interpolate, impute!
+`
+# Usage of the `dims` keyword argument
+
+julia> x = [1.0 missing; missing 2.0]
+2×2 Matrix{Union{Missing, Float64}}:
+ 1.0        missing
+  missing  2.0
+
+julia> imp = Substitute(statistic = mean)
+Substitute{typeof(mean)}(Statistics.mean)
+
+# Variables in columns, each row is an observation (like a DataFrame)
+julia> impute(x, imp, dims=:cols)
+2×2 Matrix{Union{Missing, Float64}}:
+ 1.0  2.0
+ 1.0  2.0
+
+# Variables in rows, each column is an observation
+julia> impute(x, imp, dims=:rows)
+2×2 Matrix{Union{Missing, Float64}}:
+ 1.0  1.0
+ 2.0  2.0
+
+# Impute over all dimensions
+julia> impute(x, imp, dims=:)
+2×2 Matrix{Union{Missing, Float64}}:
+ 1.0  1.5
+ 1.5  2.0
+
+# Default is `dims = :cols`
+julia> impute(x, imp)
+2×2 Matrix{Union{Missing, Float64}}:
+ 1.0  2.0
+ 1.0  2.0
+
+# In-place imputation
 
 julia> M = [1.0 2.0 missing missing 5.0; 1.1 2.2 3.3 missing 5.5]
 2×5 Matrix{Union{Missing, Float64}}:
  1.0  2.0   missing  missing  5.0
  1.1  2.2  3.3       missing  5.5
 
-julia> impute!(M, Interpolate(); dims=1)
```

Comment:
I think showing that `dims` can still be an integer dimension is important, as that's how it works in Base. For example, you can impute along `dims=3` for a n-dimensional array.

https://github.com/invenia/Impute.jl/blob/master/test/testutils.jl#L224